netaddr.go

```diff
@@ -1066,13 +1060,19 @@ func (p IPPrefix) Contains(ip IP) bool {
 		return false
 	}
 	if ip.Is4() {
-		m := mask4(p.Bits)
-		return uint32(ip.addr.lo)&m == uint32(p.IP.addr.lo)&m
+		// xor the IP addresses together; mismatched bits are now ones.
+		// Shift away the number of bits we don't care about.
+		// We know that p.Bits is in the range 0..32 because p is Valid;
+		// the compiler doesn't know that, so mask with 63 to help it.
+		// Now truncate to 32 bits, because this is IPv4.
+		// If all the bits we care about are equal, the result wil be zero.
+		return uint32((ip.addr.lo^p.IP.addr.lo)>>((32-p.Bits)&63)) == 0
 	} else {
+		// xor the IP addresses together.
+		// Mask away the bits we don't care about.
+		// If all the bits we care about are equal, the result wil be zero.
 		m := mask6(p.Bits)
-		// TODO: benchmark whether the short circuit below is faster or slower
-		// than the higher level alternative: 'ip.addr.and(m) == p.IP.addr.and(m)'.
-		return ip.addr.hi&m.hi == p.IP.addr.hi&m.hi && ip.addr.lo&m.lo == p.IP.addr.lo&m.lo
+		return ip.addr.xor(p.IP.addr).and(m).isZero()
```

Comment:
Vaguely surprised that you can't make it faster using the same tricks as v4. I would have expected it to scale fairly well to two parallel uint64 operations on OOO hardware. But, benchmarks don't lie I guess.

Reply:
The trouble is implementing uint128.shr efficiently. You need a branch, and an extra shift and or.

In the last hour I’ve started contemplating a proposal to add new API to math/bits, something like:

```go
// ShiftInRight64 right shifts lo by n bits.
// The lower n bits of hi are shifted in to replace them.
// If n is greater than 64, any extra bits shifted in are all zero.
func ShiftInRight64(hi, lo uint64, n uint8) uint64
```

Plus obvious friends to fill out the family (different bit widths, ShiftInLeft).

This would correspond to amd64’s SHRD instruction, plus some fix-up code for large shifts.

I don’t know what n’s type should be. int, uint, and uint8 are all natural choices.

math/big could then use this to implement its shrVU_g and shlVU_g primitives, maybe efficiently enough to allow us to delete the amd64 assembly.